SecurityException for getComponentDrawable

Got some automated reports with this stack trace for v3.16.0, which should have included all the latest fixes?

@TBog if you can have a look? Sorry to keep bothering you, but this icon change definitely had a larger impact than expected!

```
java.lang.RuntimeException: 
  at android.os.AsyncTask$4.done (AsyncTask.java:399)
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:383)
  at java.util.concurrent.FutureTask.setException (FutureTask.java:252)
  at java.util.concurrent.FutureTask.run (FutureTask.java:271)
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:289)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at java.lang.Thread.run (Thread.java:919)
Caused by: java.lang.SecurityException: 
  at android.os.Parcel.createException (Parcel.java:2088)
  at android.os.Parcel.readException (Parcel.java:2056)
  at android.os.Parcel.readException (Parcel.java:2004)
  at android.content.pm.IPackageManager$Stub$Proxy.getApplicationInfo (IPackageManager.java:5785)
  at android.app.ApplicationPackageManager.getApplicationInfoAsUser (ApplicationPackageManager.java:487)
  at android.app.ApplicationPackageManager.getLiveIcon (ApplicationPackageManager.java:4275)
  at android.app.ApplicationPackageManager.loadUnbadgedItemIcon (ApplicationPackageManager.java:3483)
  at android.app.ApplicationPackageManager.loadUnbadgedItemIcon (ApplicationPackageManager.java:3429)
  at android.app.ApplicationPackageManager.loadItemIcon (ApplicationPackageManager.java:3323)
  at android.content.pm.PackageItemInfo.loadIcon (PackageItemInfo.java:310)
  at android.content.pm.PackageItemInfo.loadIcon (PackageItemInfo.java:303)
  at android.content.pm.LauncherActivityInfo.getIcon (LauncherActivityInfo.java:143)
  at android.content.pm.LauncherActivityInfo.getIcon (LauncherActivityInfo.java:118)
  at android.content.pm.LauncherActivityInfo.getBadgedIcon (LauncherActivityInfo.java:199)
  at fr.neamar.kiss.icons.SystemIconPack.getComponentDrawable (SystemIconPack.java:86)
  at fr.neamar.kiss.IconsHandler.getDrawableIconForPackage (IconsHandler.java:170)
  at fr.neamar.kiss.result.AppResult.getDrawable (AppResult.java:443)
  at fr.neamar.kiss.result.Result$AsyncSetImage.doInBackground (Result.java:441)
  at fr.neamar.kiss.result.Result$AsyncSetImage.doInBackground (Result.java:419)
  at android.os.AsyncTask$3.call (AsyncTask.java:378)
  at java.util.concurrent.FutureTask.run (FutureTask.java:266)
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:289)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at java.lang.Thread.run (Thread.java:919)
Caused by: android.os.RemoteException: 
  at android.app.ContextImpl.enforce (ContextImpl.java:2002)
  at android.app.ContextImpl.enforceCallingOrSelfPermission (ContextImpl.java:2030)
  at com.android.server.pm.permission.PermissionManagerService.enforceCrossUserPermission (PermissionManagerService.java:3879)
  at com.android.server.pm.permission.PermissionManagerService.access$2500 (PermissionManagerService.java:157)
  at com.android.server.pm.permission.PermissionManagerService$PermissionManagerServiceInternalImpl.enforceCrossUserPermission (PermissionManagerService.java:4135)
```

--- a/app/src/main/java/fr/neamar/kiss/IconsHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/IconsHandler.java
@@ -150,7 +150,7 @@ public class IconsHandler {
         }
 
         // check the icon pack for a resource
-        if (mIconPack != null) {
+        if (mIconPack != null && userHandle.isCurrentUser()) {
             // just checking will make this thread wait for the icon pack to load
             if (!mIconPack.isLoaded())
                 return null;
@@ -174,7 +174,7 @@ public class IconsHandler {
             return null;
 
         // if the icon pack has a mask, use that instead of the adaptive shape
-        if (mIconPack != null && mIconPack.hasMask()) {
+        if (mIconPack != null && mIconPack.hasMask() && userHandle.isCurrentUser()) {
             Drawable drawable = mIconPack.applyBackgroundAndMask(ctx, systemIcon, false);
             storeDrawable(cacheGetFileName(cacheKey), drawable);
             return drawable;

--- a/app/src/main/java/fr/neamar/kiss/IconsHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/IconsHandler.java
@@ -29,6 +29,7 @@ import java.util.List;
 import fr.neamar.kiss.icons.IconPack;
 import fr.neamar.kiss.icons.IconPackXML;
 import fr.neamar.kiss.icons.SystemIconPack;
+import fr.neamar.kiss.pojo.AppPojo;
 import fr.neamar.kiss.result.AppResult;
 import fr.neamar.kiss.utils.DrawableUtils;
 import fr.neamar.kiss.utils.UserHandle;
@@ -139,10 +140,11 @@ public class IconsHandler {
     @SuppressWarnings("CatchAndPrintStackTrace")
     public Drawable getDrawableIconForPackage(ComponentName componentName, UserHandle userHandle) {
         final String componentString = componentName.toString();
+        final String cacheKey = AppPojo.getComponentName(componentName.getPackageName(), componentName.getClassName(), userHandle);
 
         // Search in cache
         {
-            Drawable cacheIcon = cacheGetDrawable(componentString);
+            Drawable cacheIcon = cacheGetDrawable(cacheKey);
             if (cacheIcon != null)
                 return cacheIcon;
         }
@@ -161,7 +163,7 @@ public class IconsHandler {
                     drawable = DrawableUtils.applyIconMaskShape(ctx, iconPackDrawable, shape, true);
                 } else
                     drawable = mIconPack.applyBackgroundAndMask(ctx, iconPackDrawable, false);
-                storeDrawable(cacheGetFileName(componentString), drawable);
+                storeDrawable(cacheGetFileName(cacheKey), drawable);
                 return drawable;
             }
         }
@@ -174,7 +176,7 @@ public class IconsHandler {
         // if the icon pack has a mask, use that instead of the adaptive shape
         if (mIconPack != null && mIconPack.hasMask()) {
             Drawable drawable = mIconPack.applyBackgroundAndMask(ctx, systemIcon, false);
-            storeDrawable(cacheGetFileName(componentString), drawable);
+            storeDrawable(cacheGetFileName(cacheKey), drawable);
             return drawable;
         }
 
@@ -187,7 +189,7 @@ public class IconsHandler {
         else
             drawable = systemIcon;
 
-        storeDrawable(cacheGetFileName(componentString), drawable);
+        storeDrawable(cacheGetFileName(cacheKey), drawable);
         return drawable;
     }
 

--- a/app/src/main/java/fr/neamar/kiss/icons/SystemIconPack.java
+++ b/app/src/main/java/fr/neamar/kiss/icons/SystemIconPack.java
@@ -5,19 +5,9 @@ import android.content.Context;
 import android.content.pm.LauncherActivityInfo;
 import android.content.pm.LauncherApps;
 import android.content.pm.PackageManager;
-import android.graphics.Bitmap;
-import android.graphics.BitmapShader;
-import android.graphics.Canvas;
-import android.graphics.Color;
-import android.graphics.Paint;
-import android.graphics.Shader;
-import android.graphics.drawable.AdaptiveIconDrawable;
-import android.graphics.drawable.BitmapDrawable;
 import android.graphics.drawable.Drawable;
-import android.graphics.drawable.LayerDrawable;
 import android.os.Build;
 import android.util.Log;
-import android.util.TypedValue;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
@@ -83,19 +73,28 @@ public class SystemIconPack implements IconPack<Void> {
                 List<LauncherActivityInfo> icons = launcher.getActivityList(componentName.getPackageName(), userHandle.getRealHandle());
                 for (LauncherActivityInfo info : icons) {
                     if (info.getComponentName().equals(componentName)) {
-                        drawable = info.getBadgedIcon(0);
-                        break;
+                        try {
+                            drawable = info.getBadgedIcon(0);
+                            break;
+                        } catch (SecurityException ignored) {
+                            // https://github.com/Neamar/KISS/issues/1715
+                            // not sure how to avoid it so we catch and ignore
+                        }
                     }
                 }
 
-                // This should never happen, let's just return the first icon
+                // This should never happen, let's just return the activity icon
                 if (drawable == null)
-                    drawable = icons.get(0).getBadgedIcon(0);
+                    drawable = ctx.getPackageManager().getActivityIcon(componentName);
             } else {
                 drawable = ctx.getPackageManager().getActivityIcon(componentName);
             }
+
+            // This should never happen, let's just return the generic activity icon
+            if (drawable == null)
+                drawable = ctx.getPackageManager().getDefaultActivityIcon();
         } catch (PackageManager.NameNotFoundException | IndexOutOfBoundsException e) {
-            Log.e(TAG, "Unable to find component " + componentName.toString() + e);
+            Log.e(TAG, "Unable to find component " + componentName.toShortString(), e);
         }
         return drawable;
     }

